From: Oleksij Rempel <o.rempel@pengutronix.de>
Date: Thu, 14 Mar 2024 10:09:12 +0100
Subject: [PATCH] ARM: i.MX8MP: skov: add fallback for unsupported HW

Add fallback support for currently not supported HW. Instead of
completely fail, it is better to use some kind of basic devicetree with
reduced functionality.

Signed-off-by: Oleksij Rempel <o.rempel@pengutronix.de>
---
 arch/arm/boards/skov-imx8mp/board.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boards/skov-imx8mp/board.c b/arch/arm/boards/skov-imx8mp/board.c
index 5424c3b9ab5f..a45fe8517f59 100644
--- a/arch/arm/boards/skov-imx8mp/board.c
+++ b/arch/arm/boards/skov-imx8mp/board.c
@@ -83,6 +83,12 @@ static const struct board_description imx8mp_variants[] = {
 		.dts_compatible_hdmi = "skov,imx8mp-skov-revb-hdmi",
 		.flags = SKOV_IMX8MP_HAS_HDMI,
 	},
+	/* Keep this entry always on the last place. It will be used
+	 * as fallback variant if not supported variant is detected.
+	 */
+	{
+		.dts_compatible = "skov,imx8mp-skov-basic",
+	},
 };
 
 static int skov_imx8mp_fixup(struct device_node *root, void *data)
@@ -168,13 +174,14 @@ static int skov_imx8mp_init_variant(struct skov_imx8mp_priv *priv)
 	if (ret)
 		return ret;
 
-	priv->variant_id = v;
 
 	if (v >= ARRAY_SIZE(imx8mp_variants)) {
-		dev_err(dev, "Invalid variant %u\n", v);
-		return -EINVAL;
+		dev_warn(dev, "Unsuported variant %u. Fall back to basic variant\n", v);
+		v = ARRAY_SIZE(imx8mp_variants) - 1;
 	}
 
+	priv->variant_id = v;
+
 	variant = &imx8mp_variants[v];
 
 	if (variant->flags & SKOV_IMX8MP_HAS_HDMI) {
